src/main/java/io/github/netmikey/mvncloner/mvncloner/Publisher.java

```diff
@@ -39,63 +39,86 @@ public class Publisher {
                 }
             }
         }
 
+        for (var future : futures) {
+            future.get();
+        }
+
         // Tail recursion
         for (Path recursePath : recursePaths) {
             String subpath = mirrorPath.relativize(recursePath).toString();
             publishDirectory(httpClient, appendUrlPathSegment(repositoryUrl, subpath), recursePath);
         }
     }
 
-    private void handleFile(HttpClient httpClient, String repositoryUrl, Path path)
-        throws IOException, InterruptedException {
+    private void handleFile(HttpClient httpClient, String repositoryUrl, Path path) {
 
         String filename = path.getFileName().toString();
         String targetUrl = repositoryUrl + filename;
-        LOG.info("Uploading " + targetUrl);
-
-        Utils.sleep(1000);
-        HttpRequest request = Utils.setCredentials(HttpRequest.newBuilder(), username, password)
-            .uri(URI.create(targetUrl))
-            .PUT(BodyPublishers.ofInputStream(() -> {
-                try {
-                    return Files.newInputStream(path, StandardOpenOption.READ);
-                } catch (IOException e) {
-                    throw new RuntimeException(e);
-                }
-            }))
-            .build();
-        HttpResponse<String> response = httpClient.send(request, BodyHandlers.ofString());
-        if (response.statusCode() < 200 || response.statusCode() > 299) {
-            LOG.error("Error uploading " + targetUrl + " : Response code was " + response.statusCode());
-            LOG.debug("   Response headers: " + response.headers());
-            LOG.debug("   Response body: " + response.body());
+        try {
+            var baseReq = Utils.setCredentials(HttpRequest.newBuilder(), username, password)
+                .uri(URI.create(targetUrl));
+
+            var getResponse = httpClient.send(baseReq.method("HEAD", BodyPublishers.noBody()).build(), BodyHandlers.discarding());
+            if (getResponse.statusCode() != 404) {
+                LOG.info("Artifact {} already exists", targetUrl);
+                LOG.debug("   Response headers: " + getResponse.headers());
+                return;
+            }
+
+            LOG.info("Uploading " + targetUrl);
+            Utils.sleep(uploadInterval);
+            if (dryRun) return;
+
+            var putRequest = baseReq.PUT(BodyPublishers.ofFile(path)).build();
+
+            HttpResponse<String> putResponse = httpClient.send(putRequest, BodyHandlers.ofString());
+            if (putResponse.statusCode() < 200 || putResponse.statusCode() > 299) {
+                LOG.error("Error uploading " + targetUrl + " : Response code was " + putResponse.statusCode());
+                LOG.debug("   Response headers: " + putResponse.headers());
+                LOG.debug("   Response body: " + putResponse.body());
+            }
+        } catch (Exception e) {
+            LOG.error("Fail to send " + filename + " to " + targetUrl, e);
```

Comment:
This changes the behavior quite a bit: Before, the Publisher would abort on `IOException` and `InterruptedException` (which probably won't be recoverable anyway, so no use to continue to try uploading). Now, you just log and continue uploading.

I'd rather not change the general error behavior, so if you'd like to introduce concurrency, please handle errors accordingly.

Reply:
Ok I see.

This seems legit for `IOException` then, but the tool should definitely always abort on `InterruptedException`.

Thinking further here: I'd like to avoid having to scrape the logs at the end of a long upload session to find if everything worked or not. If the tool aborts at the first error, it is clear where / when things went wrong but if the tool continues for a long time, errors might get overseen.

I can think of 2 things to help solve this:

1. Make it configurable: something like `--target.abort-on-error=false`. This way the old behavior can be restored for people on stable networks that might prefer the current behavior.
2. Count the number of failed uploads and print a summary line at the very end (in addition to the error logs). This way, if `--target.abort-on-error=false` is set, you can be sure that nothing failed, or how much failed during the long-running operation.

Reply:
collect erros recursively could be more complex, we need to think more about lifecycle of Publisher instance, usually we re-run to check if some fail has ocurred.

But the flag is a good thing, i'll add it